### PR TITLE
Prompt embeddings

### PR DESCRIPTION
**Pull request by rhendz**

### CODE REVIEW

File: spice_agent/inference/actions.py [hunk 1]
```diff
@@ -21,12 +23,156 @@
...
+            pipe,
+            (
+                StableDiffusionPipeline,
+                StableDiffusionXLPipeline,
+                StableDiffusionXLImg2ImgPipeline,
+            ),
+        ):
+            message = f"Prompt embeddings are not supported for pipeline {type(pipe)}"
+            LOGGER.warn(message)
+            return None
```
Comment (djstein) [inline spice_agent/inference/actions.py]: should this be done before before making an instance of this class?

File: spice_agent/inference/actions.py [hunk 2]
```diff
@@ -21,12 +23,156 @@
...
+        Initializes compel text embedding system based on type of pipeline.
+        """
+
+        # Compel settings consistent across compel objects
+        DEFAULT_COMPEL_ARGUMENTS = {
+            "truncate_long_prompts": False,
+            "device": self.device,
+        }
+
+        pipe = self.pipeline
```
Comment (djstein) [inline spice_agent/inference/actions.py]: no need to rename

File: spice_agent/inference/actions.py (hunk 1 above)
Comment (djstein) [reply spice_agent/inference/actions.py]: might be okay

File: spice_agent/inference/actions.py [hunk 3]
```diff
@@ -21,12 +23,156 @@
...
+                **DEFAULT_COMPEL_ARGUMENTS,
+            )
+
+        # Configure compel objects for StableDiffusionXL pipelines
+        if isinstance(pipe, StableDiffusionXLPipeline):
+            return Compel(
+                tokenizer=[pipe.tokenizer, pipe.tokenizer_2],
+                text_encoder=[pipe.text_encoder, pipe.text_encoder_2],
+                returned_embeddings_type=ReturnedEmbeddingsType.PENULTIMATE_HIDDEN_STATES_NON_NORMALIZED,
+                requires_pooled=[False, True],
```
Comment (djstein) [inline spice_agent/inference/actions.py]: different value from line 90?

File: spice_agent/inference/actions.py [hunk 4]
```diff
@@ -21,12 +23,156 @@
...
+
+        self.without_embeddings = {
+            "prompt": self.prompt,
+            "negative_prompt": self.negative_prompt,
+        }
+
+        if self.compel:
+            try:
+                return self._run_compel()
+            except Exception as exception:
```
Comment (djstein) [inline spice_agent/inference/actions.py]: maybe this should be a more explicit exception

File: spice_agent/inference/actions.py [hunk 5]
```diff
@@ -185,7 +331,16 @@ def run_pipeline(
             elif is_text_input and is_file_output:
                 SPICE_INFERENCE_DIRECTORY.mkdir(parents=True, exist_ok=True)
                 save_at = Path(SPICE_INFERENCE_DIRECTORY / f"{inference_job_id}.png")
+
                 stable_diffusion_options = self._get_stable_diffusion_options(options)
+                prompt = text_input
+                negative_prompt = ""
+                if "negative_prompt" in stable_diffusion_options:
+                    negative_prompt = stable_diffusion_options["negative_prompt"]
```
Comment (djstein) [inline spice_agent/inference/actions.py]: may be useful to use walrus
```suggestion
                if negative_prompt := stable_diffusion_options['negative_prompt"]:
```

File: spice_agent/inference/actions.py [hunk 6]
```diff
@@ -174,7 +320,7 @@ def run_pipeline(
         try:
             if is_text_input and is_text_output:
                 result = None
-                pipe = pipeline(model=model_repo_id, device=self.device)
+                pipe = transformers.pipeline(model=model_repo_id, device=self.device)
```
Comment (djstein) [inline spice_agent/inference/actions.py:323]: `pipe` will be redefined on 346

File: spice_agent/inference/actions.py [hunk 7]
```diff
@@ -210,23 +378,26 @@ def run_pipeline(
...
-                            output_type="latent",
-                            **stable_diffusion_options,
-                        ).images  # type: ignore
+                        # Generates additional embeddings for refiner
+                        prompt_embeddings_for_refiner = CompelEmbeddingsProvider(
+                            pipeline=refiner,
+                            device=self.device,
+                            prompt=prompt,
+                            negative_prompt=negative_prompt,
+                        )()
```
Comment (djstein) [inline spice_agent/inference/actions.py]: not my favorite to see a call get instantiated then immediately called

"what" got called is not clear

File: spice_agent/inference/actions.py (hunk 3 above)
Comment (rhendz) [reply spice_agent/inference/actions.py]: yep that's fine - different embeddings type are required here

File: spice_agent/inference/actions.py (hunk 4 above)
Comment (rhendz) [reply spice_agent/inference/actions.py]: I really don't expect an exception to occur here so I still catch a generic exception to log it

**Pull request closed without merging**